apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_subs.erl

```diff
@@ -169,30 +260,90 @@ on_unsubscribe(SessionId, TopicFilter, S0) ->
+            S = delete_subscription(TopicFilter, S0),
             ?tp(persistent_session_ds_subscription_delete, #{
-                session_id => SessionId, topic_filter => TopicFilter
+                session_id => SessionId,
+                topic_filter => TopicFilter,
+                mode => Mode
             }),
-            ?tp_span(
-                persistent_session_ds_subscription_route_delete,
-                #{session_id => SessionId, topic_filter => TopicFilter},
-                ok = emqx_persistent_session_ds_router:do_delete_route(TopicFilter, SessionId)
-            ),
-            _ = emqx_external_broker:delete_persistent_route(TopicFilter, SessionId),
-            S = emqx_persistent_session_ds_state:del_subscription(TopicFilter, S0),
+            %% Immediate side-effects:
+            ok = delete_route(Mode, SessionId, TopicFilter),
             {ok, gc(S), Subscription}
     end.
 
+delete_subscription(TopicFilter, S0) ->
+    emqx_persistent_session_ds_state:del_subscription(TopicFilter, S0).
+
+%% @doc Restart (and potentially downgrade) subscriptions during session replay.
+%% 1. Resubscribes to the broker for direct subscriptions.
+%% 2. Downgrades durable QoS0 subscriptions to direct subscriptions.
+%% Returns a list of events, which currently includes only subscription downgrades.
+%% To be called before the stream scheduler is initialized.
+-spec on_session_replay(emqx_persistent_session_ds:id(), emqx_persistent_session_ds_state:t()) ->
+    {emqx_persistent_session_ds_state:t(), [Event]}
+when
+    Event :: {downgrade, subscription_mode(), TopicFilter, subscription()},
+    TopicFilter :: emqx_persistent_session_ds:topic_filter().
+on_session_replay(SessionId, S0) ->
+    fold(
+        fun(TopicFilter, Sub0, {SAcc, Acc}) ->
+            case restart_subscription(TopicFilter, Sub0, SessionId, SAcc) of
+                ok ->
+                    {SAcc, Acc};
+                {Outcome, S} ->
+                    {S, [Outcome | Acc]}
+            end
+        end,
+        {S0, []},
+        S0,
+        [direct]
+    ).
+
+restart_subscription(TopicFilter, Sub0 = #{current_state := SStateId}, SessionId, S0) ->
+    SState =
+        #{subopts := SubOpts} =
+        emqx_persistent_session_ds_state:get_subscription_state(SStateId, S0),
+    DesiredMode = desired_subscription_mode(SubOpts),
+    case sstate_subscription_mode(SState) of
+        durable when DesiredMode =:= direct ->
+            %% Downgrade from durable to direct is now possible:
+            S = downgrade_subscription(TopicFilter, SState, SessionId, S0),
+            {{downgrade, DesiredMode, TopicFilter, Sub0}, S};
+        direct ->
+            %% Restore the subscription in the broker:
+            ok = add_route(direct, SessionId, TopicFilter, SubOpts);
+        durable ->
+            %% Nothing to do:
+            ok
+    end.
+
+downgrade_subscription(TopicFilter, SState, SessionId, S0) ->
```

Comment:
`downgrade_subscription` -> `change_to_direct`

Terms `upgrade`/`downgrade` are ambiguous, and may collide with possible future usecases where the state should be updated according to the code change.

Reply:
Changed in [5c22ce5](https://github.com/emqx/emqx/pull/15261/commits/5c22ce53578c0f52e7e9c595776816d8f36c9781).